### PR TITLE
Check input values

### LINKED ISSUE #302: Assert that entries to softmax are appropriate
No single entry in the input array of the softmax should exceed that shape of the input array.

So for example,  if you call `F.softmax(y)` then `assert y.max() < y.shape[1]` should be True. This does introduce some overhead, so perhaps is worth debating -- but I just spent a few hours debugging such a bug so I'm in favor :)

### PR DESCRIPTION
**Pull request by unnonouno**
I'll add check code to functions
- array boundary
- value condition for math functions (such as logarithm)

fix #302 
related to #457 

Please merge #899 before this PR

### CODE REVIEW

File: chainer/__init__.py [hunk 1]
```diff
@@ -29,4 +29,26 @@
...
 AUTO = flag.AUTO
 
+_debug = False
+
+
+def get_debug():
+    """Get the debug mode.
+
+    Returns:
+        bool: Return True if Chainer is in debug mode.
```
Comment (delta2323) [inline chainer/__init__.py]: ``True``

File: chainer/functions/array/select_item.py [hunk 2]
```diff
@@ -22,21 +23,26 @@ def check_type_forward(self, in_types):
...
-            'S t, raw T x',
-            'T y',
-            'int ind[] = {i, t}; y = x[ind];',
-            'getitem_fwd'
-        )(t, x)
-        return y,
+        if chainer.get_debug():
+            if not ((0 <= t).all() and
+                    (t < x.data.shape[1]).all()):
+                raise RuntimeError
```
Comment (delta2323) [inline chainer/functions/array/select_item.py]: Add message to `RuntimeError`.

File: chainer/functions/connection/embed_id.py [hunk 3]
```diff
@@ -22,6 +23,11 @@ def check_type_forward(self, in_types):
 
     def forward(self, inputs):
         x, W = inputs
+        if chainer.get_debug():
+            if not ((0 <= x).all() and
+                    (x < W.data.shape[0]).all()):
+                raise RuntimeError
```
Comment (delta2323) [inline chainer/functions/connection/embed_id.py]: ditto(Add error message)

File: chainer/functions/connection/embed_id.py [hunk 4]
```diff
@@ -22,6 +23,11 @@ def check_type_forward(self, in_types):
 
     def forward(self, inputs):
         x, W = inputs
+        if chainer.get_debug():
+            if not ((0 <= x).all() and
+                    (x < W.data.shape[0]).all()):
```
Comment (delta2323) [inline chainer/functions/connection/embed_id.py]: Use `len(W.data)` instead.

File: chainer/functions/loss/softmax_cross_entropy.py [hunk 5]
```diff
@@ -67,8 +68,16 @@ def check_type_forward(self, in_types):
             x_type.shape[2:] == t_type.shape[1:],
         )
 
+    def _check_input_values(x, t):
+        if not ((0 <= t).all() and
+                (t < x.data.shape[1]).all()):
+            raise RuntimeError
```
Comment (delta2323) [inline chainer/functions/loss/softmax_cross_entropy.py]: ditto (add error message).

File: tests/chainer_tests/test_function.py [hunk 6]
```diff
@@ -316,4 +316,50 @@ def test_forward_invalid(self):
...
+class TestFunctionDebug(unittest.TestCase):
+
+    def setUp(self):
+        chainer.set_debug(True)
+        self.one = numpy.array([1], numpy.float32)
+        self.nan = numpy.array([float('nan')], numpy.float32)
+        self.f = chainer.Function()
+
+    def tearDown(self):
+        chainer.set_debug(False)
```
Comment (delta2323) [inline tests/chainer_tests/test_function.py]: Original debug mode not necessarily `False`. So please store current debug mode in `setUp` and restore it in `tearDown`.

File: tests/chainer_tests/test_function.py [hunk 7]
```diff
@@ -316,4 +316,50 @@ def test_forward_invalid(self):
             f(v)
 
 
+class TestFunctionDebug(unittest.TestCase):
```
Comment (delta2323) [inline tests/chainer_tests/test_function.py]: It is desirable if we can add test cases in which we check functions do not raise error when debug mode if `False`.

**Comment by delta2323** [comment]
LGTM :+1:
